nni/algorithms/compression/v2/pytorch/utils/external/huggingface.py

```diff
@@ -0,0 +1,141 @@
+            if pattern in module_name:
+                return True
+        return False
+
+    @classmethod
+    def is_ffn(cls, module_name: str, ffn_num: int = 1) -> bool:
+        if cls.is_attention(module_name):
+            return False
+        if ffn_num == 1:
+            for pattern in cls.FFN1:
+                if pattern in module_name:
+                    return True
+        if ffn_num == 2:
+            for pattern in cls.FFN2:
+                if pattern in module_name:
+                    return True
+        return False
+
+    @classmethod
+    def get_num_heads(cls, module_name: str, model: Module) -> int:
+        if cls.is_attention(module_name, include_output=True):
+            for pattern in cls.ATTENTION:
+                match = re.search(pattern, module_name)
+                if match:
+                    attention_module_name = module_name[0: match.span()[1]]
+                    module = get_nested_attr(model, attention_module_name)
+                    if hasattr(module, 'num_attention_heads'):
+                        num_heads = module.num_attention_heads
+                    elif hasattr(module, 'num_heads'):
+                        num_heads = module.num_heads
+                    elif hasattr(module, 'n_heads'):
+                        num_heads = module.n_heads
+                    else:
+                        warn_msg = f'Can not get the heads number of attention layer : {attention_module_name}.'
+                        _logger.warning(warn_msg)
+                        num_heads = 0
+                    return num_heads
+            return 0
+        else:
+            warn_msg = f'The layer `{module_name}` might not an (Q|K|V) attention layer.'
+            _logger.warning(warn_msg)
+            return 0
+
+
+class HuggingfaceBertParser(HuggingfaceModelParser):
+    TRANSFORMER_PREFIX = r'bert\.encoder\.layer\.[0-9]+\.'
+    QKV = ('attention.self.query', 'attention.self.key', 'attention.self.value')
+    QKVO = QKV + ('attention.output.dense',)
+    FFN1 = ('intermediate.dense',)
+    FFN2 = ('output.dense',)
+    ATTENTION = ('attention.self',)
+
+
+class HuggingfaceBartParser(HuggingfaceModelParser):
+    TRANSFORMER_PREFIX = r'(en|de)coder\.layer\.[0-9]+\.'
+    QKV = ('self_attn.q_proj', 'self_attn.k_proj', 'self_attn.v_proj', 'encoder_attn.q_proj', 'encoder_attn.k_proj', 'encoder_attn.v_proj')
+    QKVO = QKV + ('self_attn.out_proj', 'encoder_attn.out_proj')
+    FFN1 = ('fc1',)
+    FFN2 = ('fc2',)
+    ATTENTION = ('self_attn', 'encoder_attn')
+
+
+class HuggingfaceT5Parser(HuggingfaceModelParser):
+    TRANSFORMER_PREFIX = r'(en|de)coder\.block\.[0-9]+\.layer\.[0-9]+.'
+    QKV = ('SelfAttention.q', 'SelfAttention.k', 'SelfAttention.v', 'EncDecAttention.q', 'EncDecAttention.k', 'EncDecAttention.v')
+    QKVO = QKV + ('SelfAttention.o', 'EncDecAttention.o')
+    FFN1 = ('DenseReluDense.wi',)
+    FFN2 = ('DenseReluDense.wo',)
+    ATTENTION = ('SelfAttention', 'EncDecAttention')
```

Comment:
do we plan to allow users to customize parser?

Reply:
yes, it's on planing